File: Sources/SwiftMsQuicOpenSSLUtils/CertBridge.m
//
//  CertBridge.m
//  SwiftMsQuic
//
//  Created by Gyuhwan Park on 2/4/26.
//


#import <Foundation/Foundation.h>
#import <Security/Security.h>

#include "openssl/x509.h"
#include "openssl/x509_vfy.h"
#include "openssl/crypto.h"

#import "CertBridge.h"

static NSString * const kCertBridgeErrorDomain = @"pl.unstabler.opensource.swift-msquic.utils.openssl";

typedef NS_ENUM(NSInteger, CertBridgeErrorCode) {
    CertBridgeErrorCodeNullInput = -1,
    CertBridgeErrorCodeDEREncodingFailed = -2,
    CertBridgeErrorCodeSecCertificateCreationFailed = -3,
};

@implementation CertBridge

static CFArrayRef CopyEmptyCertificateArray(void) {
    return CFArrayCreate(kCFAllocatorDefault, NULL, 0, &kCFTypeArrayCallBacks);
}

+ (SecCertificateRef) copySecCertificateFromOpenSSLX509: (const void *) x509 error: (NSError **) error {
    if (x509 == NULL) {
        if (error != NULL) {
            *error = [NSError errorWithDomain: kCertBridgeErrorDomain
                                         code: CertBridgeErrorCodeNullInput
                                     userInfo: @{ NSLocalizedDescriptionKey: @"X509 pointer is NULL" }];
        }
        return nil;
    }

    // i2d_X509: Convert X509 to DER format
    // When out is NULL, OpenSSL allocates memory and returns the pointer
    unsigned char *derData = NULL;
    int derLength = i2d_X509((X509 *)x509, &derData);

    if (derLength <= 0 || derData == NULL) {
        if (error != NULL) {
            *error = [NSError errorWithDomain: kCertBridgeErrorDomain
                                         code: CertBridgeErrorCodeDEREncodingFailed
                                     userInfo: @{ NSLocalizedDescriptionKey: @"Failed to encode X509 to DER format" }];
        }
        return nil;
    }

    // Create CFData from DER bytes
    CFDataRef cfData = CFDataCreate(kCFAllocatorDefault, derData, derLength);

    // Free the DER buffer allocated by OpenSSL
    OPENSSL_free(derData);

    if (cfData == NULL) {
        if (error != NULL) {
            *error = [NSError errorWithDomain: kCertBridgeErrorDomain
                                         code: CertBridgeErrorCodeSecCertificateCreationFailed
                                     userInfo: @{ NSLocalizedDescriptionKey: @"Failed to create CFData from DER data" }];
        }
        return nil;
    }

    // Create SecCertificateRef from DER data
    SecCertificateRef certificate = SecCertificateCreateWithData(kCFAllocatorDefault, cfData);

    // Release the CFData (SecCertificateCreateWithData copies the data)
    CFRelease(cfData);

    if (certificate == NULL) {
        if (error != NULL) {
            *error = [NSError errorWithDomain: kCertBridgeErrorDomain
                                         code: CertBridgeErrorCodeSecCertificateCreationFailed
                                     userInfo: @{ NSLocalizedDescriptionKey: @"Failed to create SecCertificateRef from DER data" }];
        }
        return nil;
    }

    return certificate;
}

+ (CFArrayRef) copySecCertificateArrayFromOpenSSLStackX509: (const void *) stackX509 error: (NSError **) error {
    if (stackX509 == NULL) {
        return CopyEmptyCertificateArray();
    }

    STACK_OF(X509) *stack = (STACK_OF(X509) *)stackX509;
    int count = sk_X509_num(stack);

    if (count <= 1) {
        return CopyEmptyCertificateArray();
    }

    // Create mutable array with SecCertificateRef callbacks
    CFMutableArrayRef array = CFArrayCreateMutable(kCFAllocatorDefault, count, &kCFTypeArrayCallBacks);

    if (array == NULL) {
        if (error != NULL) {
            *error = [NSError errorWithDomain: kCertBridgeErrorDomain
                                         code: CertBridgeErrorCodeSecCertificateCreationFailed
                                     userInfo: @{ NSLocalizedDescriptionKey: @"Failed to create CFMutableArray" }];
        }
        return nil;
    }

    for (int i = 1; i < count; i++) {
        X509 *x509 = sk_X509_value(stack, i);

        NSError *certError = nil;
        SecCertificateRef cert = [self copySecCertificateFromOpenSSLX509: x509 error: &certError];

        if (cert == NULL) {
            // Clean up on failure
            CFRelease(array);
            if (error != NULL) {
                *error = certError;
            }
            return nil;
        }

        // CFArrayAppendValue retains the object
        CFArrayAppendValue(array, cert);

        // Release our reference (array now owns it)
        CFRelease(cert);
    }

    return array;
}

+ (CFArrayRef) copySecCertificateArrayFromOpenSSLStoreContext: (const void *) storeContext error: (NSError **) error {
    if (storeContext == NULL) {
        return CopyEmptyCertificateArray();
    }

    X509_STORE_CTX *context = (X509_STORE_CTX *)storeContext;
    X509_verify_cert(context);
    
    STACK_OF(X509) *stack = X509_STORE_CTX_get0_chain(context);

    return [self copySecCertificateArrayFromOpenSSLStackX509: stack error: error];
}

@end

File: Sources/SwiftMsQuicOpenSSLUtils/CertBridge.c
//
//  CertBridge.c
//  SwiftMsQuic
//
//  Created by Gyuhwan Park on 2/4/26.
//

#include <stdlib.h>
#include <string.h>

#include "openssl/x509.h"
#include "openssl/x509_vfy.h"
#include "openssl/crypto.h"

#include "CertBridge.h"

CertBridge_DERCert CertBridge_CopyDERFromX509(const void *x509) {
    CertBridge_DERCert result = { NULL, -1 };

    if (x509 == NULL) {
        return result;
    }

    unsigned char *der = NULL;
    int len = i2d_X509((X509 *)x509, &der);

    if (len <= 0 || der == NULL) {
        return result;
    }

    // Copy to stdlib-allocated buffer so caller can free() it
    result.data = (uint8_t *)malloc((size_t)len);
    if (result.data != NULL) {
        memcpy(result.data, der, (size_t)len);
        result.length = len;
    }

    OPENSSL_free(der);
    return result;
}

CertBridge_DERChain CertBridge_CopyDERChainFromStoreContext(const void *storeContext) {
    CertBridge_DERChain result = { NULL, 0 };

    if (storeContext == NULL) {
        return result;
    }

    X509_STORE_CTX *ctx = (X509_STORE_CTX *)storeContext;
    X509_verify_cert(ctx);

    STACK_OF(X509) *stack = X509_STORE_CTX_get0_chain(ctx);
    if (stack == NULL) {
        return result;
    }

    int total = sk_X509_num(stack);
    if (total <= 1) {
        // No chain certs (only leaf or empty)
        return result;
    }

    // Skip index 0 (leaf cert)
    int chainCount = total - 1;
    result.certs = (CertBridge_DERCert *)calloc((size_t)chainCount, sizeof(CertBridge_DERCert));
    if (result.certs == NULL) {
        return result;
    }

    result.count = chainCount;
    for (int i = 0; i < chainCount; i++) {
        X509 *cert = sk_X509_value(stack, i + 1);
        result.certs[i] = CertBridge_CopyDERFromX509(cert);
    }

    return result;
}

void CertBridge_Free(void *ptr) {
    free(ptr);
}
